docs/_ext/autoclass_with_examples.py

```diff
@@ -0,0 +1,108 @@
+import re
+import sys
+from typing import Any, Dict, List, Tuple
+
+from docutils.nodes import Node
+from sphinx.ext.autodoc.directive import AutodocDirective
+
+from starknet_py.constants import ROOT_PATH
+
+
+class AutoclassWithExamples(AutodocDirective):
+    """
+    Custom extension of the AutodocDirective class, which is used to
+    add code examples to the method docstrings in the documentation.
+
+    This class runs before the AutodocDirective class and pulls code snippets
+    from the starknet_py/tests/e2e/docs/code_examples directory to include in the documentation.
+    This allows developers to easily see and understand how the methods
+    being documented are intended to be used in a practical context.
+    """
+
+    def run(self) -> List[Node]:
+        original_class = getattr(
+            sys.modules[self.env.ref_context.get("py:module")], self.arguments[0]
+        )
+        add_code_examples(original_class)
+
+        self.name = self.name.replace("-with-examples", "")  # remove `-with-examples`
+
+        return AutodocDirective.run(self)
+
+
+def add_code_examples(original_class: Any):
+    """
+    Adds code examples for the given class.
+    """
+    base_class = original_class.__base__
+
+    file_name, file_content = _extract_file_properties(original_class.__name__)
+
+    for method_name, method in original_class.__dict__.items():
+        if not callable(method):
+            continue
+
+        stripped_method_name = method_name.strip("_")
+        if _code_example_exists(stripped_method_name, file_content):
+            hint = _create_hint(file_name, stripped_method_name)
+            _append_hint(method_name, method, base_class, hint)
+
+
+def _append_hint(method_name: str, method, base_class: Any, hint: str):
+    """
+    If method does not have the __doc__, takes it from the base method.
+    """
+    if method.__doc__ is None and method_name in base_class.__dict__:
+        method.__doc__ = getattr(base_class, method_name).__doc__ + hint
+    else:
+        method.__doc__ = (method.__doc__ or "") + hint
```

Comment:
This wouldn't work if the docstring was several classes "above" in the hierarchy, right?

Reply:
Yes. Since we do not currently use deep inheritance I would not implement this

Reply:
Maybe let's at least open issue for that so we don't forget? I can imagine the situation where docs are missing because of this.

Reply:
Although, I agree it should be somehow `remembered` I don't think the issue is the best place for that. I can add a comment or include information about it in the docstring.